### ISSUE TITLE
Support for Schema for AvionDB

### ISSUE THREAD

**Issue by vasa-develop** [dappkit/aviondb#23]
A Schema is a JSON object that allows you to define the shape and content of documents and embedded documents in a collection. You can use a schema to require a specific set of fields, configure the content of a field, or to validate changes to a document based on its beginning and ending states.

This issue is for discussing various approaches to implement Schema for AvionDB. The goal here is to pool different views on:

- Why do we need Schema?
- How we can implement Schema for AvionDB? (at least a rough idea)

This issue would be resolved to create a new issue with technical specifications for implementing Schema model.

**This is an open discussion, feel free to add your comments, suggestions.**

**Comment by deanpress** [comment]
As a possible implementation I'd recommend [hapi/joi](https://github.com/hapijs/joi), example below.

It's possible to store json schemas and later convert them to joi using a package like [enjoi](https://github.com/tlivings/enjoi).

Example json to joi validation with enjoy:
```js
const Joi = require('@hapi/joi');
const Enjoi = require('enjoi');

const schema = Enjoi.schema({
    type: 'object',
    properties: {
        firstName: {
            description: 'First name.',
            type: 'string'
        },
        lastName: {
            description: 'Last name.',
            type: 'string'
        },
        age: {
            description: 'Age in years',
            type: 'integer',
            minimum: 1
        }
    },
    'required': ['firstName', 'lastName']
});

const { error, value } = schema.validate({firstName: 'John', lastName: 'Doe', age: 45});
```

**Comment by vasa-develop** [comment]
Here is a related issue from the specs repo: https://github.com/dappkit/aviondb-specs/issues/3

**Comment by vaultec81** [comment]
I love that idea @deanpress. Joi is really good for doing schemes outside of hapi. If Scheme support is worked out. Protobuf serialization can be used potentially which saves a significant amount of space. Scheme support should definitely be looked at. We need to make sure there isn't conflicts when adding/removing, or modifying the Scheme. Another issue has to do with decentralization, if we implement Schemes we have to make sure there isn't conflict because of disconnected/offline peers. The best way to avoid conflicts would be having a specific Scheme be defined during the creation of the collection, but that has potential side effects.

> How we can implement Schema for AvionDB? (at least a rough idea)

Joi is really good option. Need to make sure it can be converted into protobuf message format if we plan on using protobuf encoding to reduce storage space. Also the Joi Scheme needs to be convertible into a object so it can be stored in the oplog. (Which I am sure all can be done with some extra code)

**Comment by vasa-develop** [comment]
Thanks for sharing your views @vaultec81 @deanpress. 

We can discuss more the issues that @vaultec81 has stated in terms of conflicts. I'll work on a new issue for specs to take this forward ASAP.

**Comment by vasa-develop** [comment]
@vaultec81 Should we add the CID of the Schema with the oplog? That way the oplog would stay clean & light.

**Comment by vaultec81** [comment]
@vasa-develop I don't think we need to do that. The amount data used compared to normal entries is negligible. Having it stored on IPFS adds extra time to find and fetch, introducing the potential for the data to get lost. Given its essential it should be part of the oplog.

**Comment by vasa-develop** [comment]
@vaultec81 So are you suggesting to add the Schema JSON with every oplog? 
Also, about the Protobuf conversion, is that done on `orbit-db-io` level?

**Comment by vaultec81** [comment]
@vasa-develop each entry on the oplog would reference an ID that connects back the Scheme JSON data. Not directly storing the Scheme with every entry. Also since there is a other operations aside from inserts such as $set. The Scheme validation would happen *after* the state is calculated. 

Protobuf serialization wouldn't happen on the `orbit-db-io` level rather the level where aviondb stores its state. It *could* run on the `orbit-db-io` level, but I worry there could be unforeseen side effects.

**Comment by vasa-develop** [comment]
> oplog would reference an ID that connects back the Scheme JSON data.

@vaultec81  So, where & how should I store the Schema JSON & their respective id(s)?

**Comment by vaultec81** [comment]
> So, where & how should I store the Schema JSON & their respective id(s)?

A separate system would need to be created/custom oplog operation types specifically for defining Schemes. It would still use the same oplog as the collection in question, but processed separately. Does that make sense?

**Comment by vasa-develop** [comment]
Yeah that makes sense. We'll need to handle things like replicating the Schema oplog with the collection oplog data. Any other possible edge cases that you can think of?

**Comment by vaultec81** [comment]
We will have to experiment with how to make a system like this work, and ensure there are no significant added conflicts/drawbacks.

>So when we replicate the collection, then the Schema will also get replicated with it, right?

Correct.

**Comment by vasa-develop** [comment]
> Protobuf serialization wouldn't happen on the orbit-db-io level rather the level where aviondb stores its state.

I'm a bit confused here. Where do we store the "state"?

**Comment by vaultec81** [comment]
> I'm a bit confused here. Where do we store the "state"?

The state is essentially what the collectionIndex is. Which we store in memory which is calculated from the oplog on start.

**Comment by vasa-develop** [comment]
Got it.

**Comment by vasa-develop** [comment]
In case we add Schema support, should we allow creating a collection without a Schema?

**Comment by vaultec81** [comment]
Mongodb is specifications is meant to allow for no Schema usage. Thats the idea behind document based databases. Requiring a Schema to be present, would be mean data formats cannot change, and the fields of the documents inserted must be defined before hand.

**Comment by vasa-develop** [comment]
Gotcha. So, we'll allow no Schema usage.

**Comment by vasa-develop** [comment]
I thought a bit about the conflict issues that we discussed above.

In case of a single peer using a database:
There are no conflicts and the peer can transform its data according to the updated schema.

In case of multiple, online/offline peers using a database:
I think there would not a conflict in the final schema, as it's just an oplog and the most recent oplog for the Schema will be the latest Schema. Here the only problem that I see is the issue of data transformation, as all the peers need to transform the data according to the latest Schema.

**Comment by deanpress** [comment]
@vasa-develop On the issue of transforming data with a new Schema:

Perhaps there could be an optional "migration file" passed along with a new Schema. The migration file could be a js method that's supposed to transform each document according to the new Schema.

If a migration file is passed with a new Schema,, aviondb automatically applies the method on each document(?) to complete the transformation.

**Comment by vasa-develop** [comment]
@deanpress That should work, but I'm concerned about the situation itself. What if the database is big and the peers (working on the same database) are changing the Schema frequently?

**Comment by vaultec81** [comment]
> What if the database is big and the peers (working on the same database) are changing the Schema frequently

The Schema should never be changed rapidly like this, if ever. Schema must hold the same across the entire collection.

**Comment by vasa-develop** [comment]
> What if the database is big and the peers (working on the same database) are changing the Schema frequently

I'm asking if an attack like this is possible or not (and what could be potential gains for the attacker?)

**Comment by shanejonas** [comment]
https://json-schema.org/ should be used to define schemas. They are widely used in the space, mongo uses JSON Schema as well:

https://docs.mongodb.com/manual/core/schema-validation/

**Comment by vaultec81** [comment]
> I'm asking if an attack like this is possible or not (and what could be potential gains for the attacker?)

Whats the gain for an attacker to flood the entire collection with garbage data? If you give someone access, if should be assumed they will not try to destroy your database.

**Comment by deanpress** [comment]
> If you give someone access, it should be assumed they will not try to destroy your database.

The account used for access (depending on the auth method) could have been compromised, giving db control to bad actors. Though I suppose in that case the database will experience trouble regardless of frequent schema updates.

**Comment by vasa-develop** [comment]
@deanpress @vaultec81 any suggestions on these [json-schema validators](https://json-schema.org/implementations.html#validators)?
How do these compare with `enjoi`?

**Comment by vasa-develop** [comment]
> The account used for access (depending on the auth method) could have been compromised, giving db control to bad actors.

This is on the developer using the access controller. He has to make sure that this doesn't happen (or is not easy to compromise).

**Comment by deanpress** [comment]
@vasa-develop 
>How do these compare with enjoi?

`enjoi` is basically json-schema to hapi joi format, so it is in fact a json-schema validator.

If you'd prefer one that's on the [json-schema validators list](https://json-schema.org/implementations.html#validators) I'd recommend ajv which is comparable to hapi joi (+ enjoi): https://github.com/ajv-validator/ajv
